Command module dev

Vysares opened the pull request:
Command handling base functionality. Everything is tested on the teensy.
Added command queue size and serial timeout variables in config.hpp. 

New commands just need an entry in the Command enum (in comUtil.hpp) and an implementation in the huge switch statement (in comUtil.cpp). Commands can be easily added while keeping related commands numerically adjacent. The enum handles all the indexing for us.

Commands that affect the command handling module itself are dubbed meta-commands, they are not added to the queue, but are instead executed immediately, this way meta-commands can manipulate the behavior of the queue without affecting themselves.

Vysares committed:
initial commit

added comUtil hpp and cpp files

Vysares committed:
Merge branch 'main' into command_module_dev

Vysares committed:
Update ComUtil hpp and cpp

First pass on command enum. Implemented getCommand(), which reads commands from serial

Vysares committed:
Completed command handling framework

Command handling base functionality complete and tested on the teensy. Added command queue size and serial timeout variables in config.hpp. Most commands still need to be added.

Vysares committed:
Update config.hpp

added comment for serial timout variable

Vysares committed:
Update comUtil.cpp

Updated description at top of comUtil.cpp

jaretanderson commented on FSW/src/util/comUtil.cpp:
```diff
@@ -40,28 +40,49 @@ int command[commandQueueSize];
-        int serialInput = Serial.parseInt();
-        return serialInput;
+        int serialInput = Serial.parseInt(); // read the command
+        if ((serialInput <= static_cast<int>(Command::DoNothing)) && (serialInput >= 0)) // check if command is valid
```
I like this logic, however would it potentially be more useful to split up these prompts? Right now the same error is printed regardless of if the check fails due to the command code being larger than `DO_NOTHING` or if it fails due to a negative `serialInput`

Something like "Command failed due to invalid serial input" separate from "Command code outside enum bounds"

Vysares replied:
Good call. Changed to print out the value of the invalid command and the range of acceptable commands.

jaretanderson reviewed (approved):
Great modularized design, and I like the syntax changes of capitalizing the enum as constants.

JashanChopra commented on FSW/src/util/comUtil.cpp:
```diff
@@ -0,0 +1,268 @@
+static int queueIndex = 0;                  // next available index of command queue
+static bool isPaused = false;               // indicates if command execution is paused
+
+// TODO: set Serial.setTimeout() in config.hpp
```
Is this TODO waiting on another feature?

Vysares replied:
Nope, good catch. This will actually go in the initialization function in main. Removed.

JashanChopra commented on FSW/src/util/comUtil.cpp:
```diff
@@ -0,0 +1,268 @@
+        int serialInput = Serial.parseInt(); // read the command
+        if ((serialInput <= static_cast<int>(Command::DO_NOTHING)) && (serialInput > 0)) // check if command is valid
+        {
+            //Serial.println("Command Recieved."); // for debugging
```
For future debugging or is this leftover from your own testing? Will you be able to change this code on flight?

JashanChopra replied:
Side comment: Will there be a telemetry module? Command received counts are a typical industry telemeter.

Vysares replied:
This is leftover from my own testing, but can be uncommented without consequence. I left it in there just in case we want it for debugging later.

Vysares replied:
The closest thing to a telemetry module is probably housekeeping. We can easily keep track of received commands since the last restart, but unless we want to store that data on the flash modules there is no way to keep track of it through a restart of the teensy.

JashanChopra commented on FSW/src/util/comUtil.cpp:
```diff
@@ -0,0 +1,268 @@
+        else
+        {
+            Serial.println("Invalid command recieved.");
+            Serial.println("(Command Handling)");
```
Is there an extra set of parentheses here or is that just for looks?

Vysares replied:
Just for looks, the parentheses get printed

JashanChopra commented on FSW/src/util/comUtil.cpp:
```diff
@@ -0,0 +1,268 @@
+            // TODO: call corresponding function in timing module
+            break;
+        
+        
```
extra newline?

Vysares replied:
Fixed

JashanChopra commented on FSW/src/util/comUtil.cpp:
```diff
@@ -0,0 +1,268 @@
+        case HEATER_OFF: 
+            // TODO: call corresponding function in houskeeping module
+            break;
+        
```
^^

Vysares committed:
Update comUtil.cpp

Updated error message for invalid commands to indicate the value of the recieved command and the range of valid commands. Fixed inconsistent newlines in switch. Removed innacurate TODO comment
